### PR TITLE
Fix #7457

### LINKED ISSUE #7457: TypeError when using both multiprocessing and gmpy
When using Sympy with both gmpy (tested on 2.0.3 and 1.5) and multiprocessing, certain operations raise a TypeError: 'argument is not an mpz'. The code works fine within a single process with gmpy or with multiple processes without gmpy but not both.

Please see code below. 

``` python
import sympy
import multiprocessing
import numpy

def thread_function(func, data, extra_kwargs=None, num_procs=None):
    if extra_kwargs:
        func = functools.partial(func, **extra_kwargs)

    if not num_procs:
        num_procs = multiprocessing.cpu_count()
    pool = multiprocessing.Pool(processes=num_procs)
    results = pool.map_async(func, data.T)
    pool.close()

    pool.join()
    return results.get()

def test_fn(data):
    x = data[0]
    y = data[1]
    circle = sympy.Circle((0,0), 10)
    line = sympy.Line(sympy.Point(0,0), sympy.Point(x,y))
    return line.intersection(circle)[0].evalf()

if __name__ == '__main__':
    data = numpy.vstack((numpy.arange(1, 100), numpy.arange(1, 100)))

    print thread_function(test_fn, data) #raises error
#    print [test_fn(data[(0,1):,i]) for i in xrange(data.shape[1])] #works fine
```

This raises a TypeError as below:

```
Exception in thread Thread-3:
Traceback (most recent call last):
  File "C:\python27\lib\threading.py", line 810, in __bootstrap_inner
    self.run()
  File "C:\python27\lib\threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "C:\python27\lib\multiprocessing\pool.py", line 376, in _handle_results
    task = get()
  File "C:\python27\lib\site-packages\sympy\geometry\point.py", line 91, in __new__
    for f in coords.atoms(Float)]))
  File "C:\python27\lib\site-packages\sympy\simplify\simplify.py", line 3839, in nsimplify
    return _real_to_rational(expr, tolerance)
  File "C:\python27\lib\site-packages\sympy\simplify\simplify.py", line 3781, in _real_to_rational
    r = nsimplify(float, rational=False)
  File "C:\python27\lib\site-packages\sympy\simplify\simplify.py", line 3861, in nsimplify
    exprval = expr.evalf(prec, chop=True)
  File "C:\python27\lib\site-packages\sympy\core\evalf.py", line 1300, in evalf
    re = C.Float._new(re, p)
  File "C:\python27\lib\site-packages\sympy\core\numbers.py", line 673, in _new
    obj._mpf_ = mpf_norm(_mpf_, _prec)
  File "C:\python27\lib\site-packages\sympy\core\numbers.py", line 56, in mpf_norm
    rv = mpf_normalize(sign, man, expt, bc, prec, rnd)
TypeError: ('argument is not an mpz', <class 'sympy.geometry.point.Point'>, (-4.47213595499958, -8.94427190999916))
```

This looks to me as if Sympy.geometry.Point is not correctly being cast into an mpz. Why this should only happen in a multiprocessing environment however, is puzzling.

**EDIT**: I'm using Sympy 0.7.5 (release), Python 2.7.6 (32-bit) and gmpy 2.0.3 (32-bit) (from binaries) on Windows 8.1 64-bit. 

**EDIT2**: This also occurs on the current Master (2-May-2014)

**EDIT3**: There was a small bug in the example code with the order of iteration of the data. This is now fixed.

### PR DESCRIPTION
**Pull request by ovolve**
Implements the fix proposed by @casevh for #7457.
I was having the same issue and this fix worked.

### CODE REVIEW

**Comment by asmeurer** [comment]
+1

File: sympy/core/numbers.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
     fnan as _mpf_nan, fzero as _mpf_zero, _normalize as mpf_normalize,
     prec_to_dps)
 from sympy.utilities.misc import debug, filldedent
+from sympy.mpmath.libmp.backend import MPZ
```
Comment (asmeurer) [inline sympy/core/numbers.py]: This needs to be just `mpmath.libmp.backend`. Mpmath isn't included with SymPy any more.

**Comment by bjodah** [comment]
In #7457 there is a snippet for reproducing the error:
```
>>> from sympy.geometry import Point
>>> pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))
```
if that is still a valid way of reproducing the issue, then perhaps that would be a nice test to add to this PR?

**Comment by asmeurer** [comment]
Yes, if this can be tested please add a test.

**Comment by asmeurer** [comment]
`pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))` works for me in master.

**Comment by cgogolin** [comment]
We can not parallelize a scientific computation because of this bug. As a (relatively simple) fix already exists, is there a change that it will be merged into master soon?

**Comment by moorepants** [comment]
It sounds like you need to add the suggested test and this will then be merged.

**Comment by moorepants** [comment]
But, Aaron claims this is already fixed in master (at least for that test).

**Comment by cgogolin** [comment]
The bug I am actually hit by is #7457. As far as I understand the discussion over there, the root of the problem is that after/during the pickling/unpickling the coordinates of the point are being converted to a Rational.
Indeed with the old version on my Debian stable I see this behavior:
```
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.geometry import Point
>>> import pickle
>>> pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))
Point(11/10, 21/10)
>>> import sympy
>>> sympy.__version__
'0.7.6'
```
With the current master I get:
```
Python 2.7.9 (default, Mar  1 2015, 12:57:24) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.geometry import Point
>>> import pickle
>>> pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))
Point2D(11/10, 21/10)
>>> import sympy
>>> sympy.__version__
'1.0.1.dev'
```
which is slightly different (Point2D instead of Point) but the coordinates are still converted to Rational.

As far as I understand the discussion above, merging this this fix would resolve #7457.

A possible test, based on the above minimal example, would be
```type(pickle.loads(pickle.dumps(Point(1.1,2.1).evalf())).x) == type(1.1)```
which, I think, should evaluate to `True`, but currently gives `False`.

Please forgive me in case I have overlooked something.

**Comment by moorepants** [comment]
`evalf()` will convert the rationals to SymPy Float not Python float, see:

```
In [1]: %paste
>>> from sympy.geometry import Point
>>> import pickle
>>> pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))

## -- End pasted text --
Out[1]: Point2D(11/10, 21/10)

In [3]: _.evalf()
Out[3]: Point2D(1.10000000000000, 2.10000000000000)

In [4]: _.x
Out[4]: 1.10000000000000

In [6]: type(_)
Out[6]: sympy.core.numbers.Float

In [7]: type(1.1)
Out[7]: float
```

**Comment by moorepants** [comment]
Also note that when you pass Python floats into the Point object that they are converted into SymPy Rational objects.

**Comment by cgogolin** [comment]
Thank you, I now understand that my proposed "test" was idiotic. Moreover, I had misunderstood the problem described in #7457. I thought that the conversion from Float to Rational was the reason for #7457. In now understand that this is intended behavior and that instead the real problem is the type error.

For further testing I ran the minimal example from #7457 that leads to the type error on the machine on which we are trying to run the numerics. On this machine I can still reproduce the type error with the current master of sympy as follows:
```
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> from sympy.geometry import Point
>>> pickle.loads(pickle.dumps(Point(1.1,2.1)))
Point2D(11/10, 21/10)
>>> pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/pickle.py", line 1382, in loads
    return Unpickler(file).load()
  File "/usr/lib/python2.7/pickle.py", line 858, in load
    dispatch[key](self)
  File "/usr/lib/python2.7/pickle.py", line 1133, in load_reduce
    value = func(*args)
  File "sympy/geometry/point.py", line 542, in __new__
    for f in coords.atoms(Float)]))
  File "sympy/simplify/simplify.py", line 1142, in nsimplify
    return _real_to_rational(expr, tolerance)
  File "sympy/simplify/simplify.py", line 1242, in _real_to_rational
    r = nsimplify(float, rational=False)
  File "sympy/simplify/simplify.py", line 1164, in nsimplify
    exprval = expr.evalf(prec, chop=True)
  File "sympy/core/evalf.py", line 1401, in evalf
    re = Float._new(re, p)
  File "sympy/core/numbers.py", line 855, in _new
    obj._mpf_ = mpf_norm(_mpf_, _prec)
  File "sympy/core/numbers.py", line 97, in mpf_norm
    rv = mpf_normalize(sign, man, expt, bc, prec, rnd)
TypeError: argument is not an mpz
>>> import sympy
>>> sympy.__version__
'1.0.1.dev'
```
As already reported above, on my personal system, in contrast, the above code works with both sympy 0.7.6 and 1.0.1.dev.

Whether the bug bites or not seems to depend on other factors, like the python version (which differs between my two systems). I suppose this is the reason for why @asmeurer didn't see the type error?

**Comment by cgogolin** [comment]
I also just checked with python 3 and sympy master and can also reproduce the type error problem  in this setting:
```
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pickle
>>> from sympy.geometry import Point
>>> pickle.loads(pickle.dumps(Point(1.1,2.1)))
Point2D(11/10, 21/10)
>>> pickle.loads(pickle.dumps(Point(1.1,2.1).evalf()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/cgogolin/src/sympy/sympy/geometry/point.py", line 542, in __new__
    for f in coords.atoms(Float)]))
  File "/home/cgogolin/src/sympy/sympy/geometry/point.py", line 542, in <listcomp>
    for f in coords.atoms(Float)]))
  File "/home/cgogolin/src/sympy/sympy/simplify/simplify.py", line 1142, in nsimplify
    return _real_to_rational(expr, tolerance)
  File "/home/cgogolin/src/sympy/sympy/simplify/simplify.py", line 1242, in _real_to_rational
    r = nsimplify(float, rational=False)
  File "/home/cgogolin/src/sympy/sympy/simplify/simplify.py", line 1164, in nsimplify
    exprval = expr.evalf(prec, chop=True)
  File "/home/cgogolin/src/sympy/sympy/core/evalf.py", line 1401, in evalf
    re = Float._new(re, p)
  File "/home/cgogolin/src/sympy/sympy/core/numbers.py", line 855, in _new
    obj._mpf_ = mpf_norm(_mpf_, _prec)
  File "/home/cgogolin/src/sympy/sympy/core/numbers.py", line 97, in mpf_norm
    rv = mpf_normalize(sign, man, expt, bc, prec, rnd)
TypeError: argument is not an mpz
>>> import sympy
>>> sympy.__version__
'1.0.1.dev'
```

**Comment by asmeurer** [comment]
It works for me. Do you have gmpy installed? That could be the source of the problem.

**Comment by cgogolin** [comment]
On the machine that shows the error I have 
```
$ pip list | grep gmpy
gmpy (1.15)
gmpy2 (2.0.2)
$ pip3 list | grep gmpy
gmpy2 (2.0.2)
```

**Comment by casevh** [comment]
I have not looked at this for a long time but IIRC, the root cause is that
when unpickling mpf instances, the mantissa must be converted to a gmpy mpz
type if gmpy/gmpy2 is used or a Python int if gmpy/gmpy2 is not used.
mpmath sets MPZ to the appropriate constructor.

https://github.com/sympy/sympy/pull/10674 looks like a correct fix but it
probably should have a test added (as discussed).

**Comment by asmeurer** [comment]
Yes, let's please add a test. I haven't heard from @ovolve, so maybe someone else can take the work and open a new PR with a test? I don't have gmpy on my computer, so I can't verify any tests (but Travis CI has a gmpy build, so it will be tested there). Is there a more direct test that can be used than the `Point` thing?

**Pull request closed without merging**